caffe2/operators/hardtanh_op.cu

```diff
@@ -0,0 +1,69 @@
+#include "caffe2/core/common_gpu.h"
+#include "caffe2/core/context_gpu.h"
+#include "caffe2/operators/hardtanh_op.h"
+
+namespace caffe2 {
+namespace {
+template <typename T>
+__global__ void HardtanhKernel(const int N, const T* X, T* Y, T min_val_, T max_val_) {
+  // Utilize naive implementation of Hardtanh
+  CUDA_1D_KERNEL_LOOP(i, N) {
+    Y[i] = X[i] > max_val_ ? max_val_ : (X[i] < min_val_ ? min_val_ : X[i]);
+  }
+}
+
+template <typename T>
+__global__ void HardtanhGradientKernel(
+    const int N,
+    const T* Y,
+    const T* dY,
+    T* dX,
+    T min_val_,
+    T max_val_) {
+  const T c = lambda_ * alpha_;
+  CUDA_1D_KERNEL_LOOP(i, N) {
+    dX[i] = Y[i] >= max_val_ ? 0 : (Y[i] <= min_val_ ? 0 : dY[i]);
+  }
+}
+} // namespace
+
+template <>
+bool HardtanhOp<float, CUDAContext>::RunOnDevice() {
+  auto& X = Input(0);
+  auto* Y = Output(0);
+  CAFFE_ENFORCE_GT(X.size(), 0);
```

Comment:
It seems there is no need to be an assert here?

Reply:
Perfect.  SELU had used one, so I'd copied it over, but I agree that it seems to enforce a tautology.